### PR TITLE
Rewrite fields with `just_name` option to their actual index names in MLT

### LINKED ISSUE #11573: MoreLikeThis Query doesn't work with `just_path` option pre 2.0
the term vector API which is now used to build MLT queries get the term vector potentially from a different field (source) than it's executed against. This will result in empty queries and should be rewritten against it's index name field

### PR DESCRIPTION
**Pull request by s1monw**
Closes #11573

### CODE REVIEW

File: src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java [hunk 1]
```diff
@@ -296,4 +314,50 @@ private void handleExclude(BooleanQuery boolQuery, MultiTermVectorsRequest likeI
             boolQuery.add(query, BooleanClause.Occur.MUST_NOT);
         }
     }
+
+    /**
+     * This class converts the actual path name to the index name if they happen to be different.
+     * This is needed if the "path" : "just_name" feature is used in mappings where paths like `person.name` are indexed
+     * into just the leave name of the path ie. in this case `name`. For this case we need to somehow map those names to
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java]: leave -> leaf

File: src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java [hunk 2]
```diff
@@ -296,4 +314,50 @@ private void handleExclude(BooleanQuery boolQuery, MultiTermVectorsRequest likeI
...
+     * the actual fields to get the right statistics from the index when we rewrite the MLT query otherwise it will rewrite against
+     * the full path name which is not present in the index at all in that case.
+     * his will result in an empty query and no results are returned
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java]: his -> This

**Comment by rjernst** [comment]
LGTM

File: src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java [hunk 3]
```diff
@@ -237,6 +240,21 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
             }
             // fetching the items with multi-termvectors API
             org.apache.lucene.index.Fields[] likeFields = fetchService.fetch(items);
+            for (int i = 0; i < likeFields.length; i++) {
```
Comment (alexksikes) [inline src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java]: This looks good, but shouldn't we build support for `just_name` in the TVs API instead?

**Comment by alexksikes** [comment]
Not sure this entirely solves it. The items have fields that should be mapped to their respective index names as well before fetching. Perhaps we should add `just_name` functionality to TVs API?

File: src/main/java/org/elasticsearch/common/lucene/search/XMoreLikeThis.java [hunk 4]
```diff
@@ -827,9 +824,11 @@ private void addTermFrequencies(Map<String, Int> termFreqMap, Terms vector) thro
                 continue;
             }
 
-            DocsEnum docs = termsEnum.docs(null, null);
-            final int freq = docs.freq();
-
+            final DocsEnum docs = termsEnum.docs(null, null);
+            int freq = 0;
+            while(docs != null && docs.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
```
Comment (jpountz) [inline src/main/java/org/elasticsearch/common/lucene/search/XMoreLikeThis.java]: TermsEnum.docs can never return null?

**Comment by s1monw** [comment]
@alexksikes the TV API is pretty much broken and I am not sure if it's fixable with that regard. I really don't know how to fix this since we might reanalyze etc.

**Comment by alexksikes** [comment]
@s1monw yes so then this is a good enough fix.

**Comment by clintongormley** [comment]
@s1monw this was autoclosed when we deleted 1.x - could you reopen against master?

**Pull request closed without merging**